### PR TITLE
Fix libblastrampoline with Ipopt_jll

### PR DESCRIPTION
**Pull request by odow**
image

### CODE REVIEW

**Comment by odow** [comment]
As a note to my future self, here's a drop in replacement for `Ipopt_jll.amplexe` that we could add to Ipopt:
```Julia
import OpenBLAS32_jll
import Ipopt_jll

function amplexe(f)
    lbt_default_libs = get(ENV, "LBT_DEFAULT_LIBS", nothing)
    if lbt_default_libs === nothing
        ENV["LBT_DEFAULT_LIBS"] = OpenBLAS32_jll.libopenblas
    end
    ret = Ipopt_jll.amplexe(f)
    if lbt_default_libs === nothing
        delete!(ENV, "LBT_DEFAULT_LIBS")
    else
        ENV["LBT_DEFAULT_LIBS"] = lbt_default_libs
    end
    return ret
end
```

File: src/AmplNLWriter.jl [hunk 1]
```diff
@@ -47,13 +48,19 @@ function call_solver(
...
+        # the BLAS library via the LBT_DEFAULT_LIBS environment variable.
+        # Provide a default in case the user doesn't set.
+        blas = OpenBLAS32_jll.libopenblas
+        solver_cmd = pipeline(
+            addenv(
                 `$(solver_path) $(nl_filename) -AMPL $(options)`,
-                stdin = stdin,
-                stdout = stdout,
-            ),
+                "LBT_DEFAULT_LIBS" => get(ENV, "LBT_DEFAULT_LIBS", blas),
```
Comment (staticfloat) [inline src/AmplNLWriter.jl:58]: If you want to get _really_ fancy, you can auto-include the LP64 libraries that are currently loaded into Julia's copy of LBT:

```
using OpenBLAS32_jll
using LinearAlgebra

BLAS.lbt_forward(OpenBLAS32_jll.libopenblas_path)
```

```
default_libs = join([lib.libname for lib in BLAS.get_config().loaded_libs if lib.interface == :lp64], ";")
```

**Comment by odow** [comment]
Aaaaand. Windows is the odd duck

https://github.com/jump-dev/AmplNLWriter.jl/actions/runs/4486368238/jobs/7888794896?pr=163

**Comment by giordano** [comment]
> As a note to my future self, here's a drop in replacement for Ipopt_jll.amplexe that we could add to Ipopt:

As a general remark, I'd warn you against manipulating `ENV` directly: that's an unsafe global state, should something fail while you're editing it you can be left in a nasty state. Instead, use [`addenv`](https://docs.julialang.org/en/v1/base/base/#Base.addenv), which also affects only sub-processes.

File: src/AmplNLWriter.jl [hunk 2]
```diff
@@ -47,13 +63,20 @@ function call_solver(
     stdout::IO,
 )
     solver.f() do solver_path
-        ret = run(
-            pipeline(
+        # Solvers like Ipopt_jll use libblastrampoline. That requires us to set
+        # the BLAS library via the LBT_DEFAULT_LIBS environment variable.
+        # Provide a default in case the user doesn't set.
+        config = LinearAlgebra.BLAS.get_config()
```
Comment (odow) [inline src/AmplNLWriter.jl:69]:
```suggestion
        config = LinearAlgebra.BLAS.lbt_get_config()
```

**Comment by odow** [comment]
> As a general remark, I'd warn you against manipulating ENV directly

Sure. I don't think we'd advise using it in anything critical. (The code here uses `addenv`.)

It's more for the occasions when I want to run the binary quickly in the REPL to test something.

File: Project.toml [hunk 3]
```diff
@@ -3,16 +3,19 @@ uuid = "7c4d4715-977e-5154-bfe0-e096adeac482"
...
-Ipopt_jll = "9cc047cb-c261-5740-88fc-0cf96f7bdcc7"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
+OpenBLAS32_jll = "656ef2d0-ae68-5445-9ca0-591084a874a2"
 
 [compat]
-Ipopt_jll = "=300.1400.400"
+Ipopt_jll = "300.1400.400"
 MathOptInterface = "1.3"
+OpenBLAS32_jll = "0.3.10"
```
Comment (odow) [inline Project.toml:13]:
```suggestion
OpenBLAS32_jll = "=0.3.10"
```

File: src/AmplNLWriter.jl [hunk 4]
```diff
@@ -5,7 +5,23 @@
 
 module AmplNLWriter
 
+import LinearAlgebra
 import MathOptInterface as MOI
+import OpenBLAS32_jll
+
+function __init__()
+    if VERSION >= v"1.8"
```
Comment (odow) [inline src/AmplNLWriter.jl:13]:
```suggestion
    if VERSION >= v"1.8" && !Sys.iswindows
```

File: src/AmplNLWriter.jl [hunk 5]
```diff
@@ -39,6 +55,15 @@ struct _DefaultSolverCommand{F} <: AbstractSolverCommand
     f::F
 end
 
+@static if VERSION < v"1.8"
```
Comment (odow) [inline src/AmplNLWriter.jl:58]:
```suggestion
@static if VERSION < v"1.8" || Sys.iswindows
```

**Comment by odow** [comment]
So all we get is an unhelpful `could not spawn setenv(`'C:\\\\Users\\\\runneradmin\\\\.julia\\\\artifacts\\\\774ef6a9c59cbea84ec6a86bf6d7a2dc39eff2e6\\\\bin\\\\ipopt.exe` error.

The complication is that we use `openblas` on Windows:
https://github.com/JuliaPackaging/Yggdrasil/blob/d3cf664e736225391b0e0dbffbec28ce20688434/C/Coin-OR/Ipopt/build_tarballs.jl#L28-L32
and LBT on others
https://github.com/JuliaPackaging/Yggdrasil/blob/d3cf664e736225391b0e0dbffbec28ce20688434/C/Coin-OR/Ipopt/build_tarballs.jl#L68-L69

But it seems like the original reason is now fixed, https://github.com/JuliaLang/julia/issues/47638, so we could try rebuilding Ipopt.

**Comment by odow** [comment]
x-ref: https://github.com/JuliaPackaging/Yggdrasil/pull/6441

File: src/AmplNLWriter.jl [hunk 6]
```diff
@@ -47,13 +72,19 @@ function call_solver(
     stdout::IO,
 )
     solver.f() do solver_path
-        ret = run(
-            pipeline(
+        # Solvers like Ipopt_jll use libblastrampoline. That requires us to set
+        # the BLAS library via the LBT_DEFAULT_LIBS environment variable.
+        # Provide a default in case the user doesn't set.
+        lbt_default_libs = get(ENV, "LBT_DEFAULT_LIBS", _get_blas_libs())
```
Comment (odow) [inline src/AmplNLWriter.jl:78]:
```suggestion
        lbt_default_libs = get(ENV, "LBT_DEFAULT_LIBS", _get_blas_libs())
        @show lbt_default_libs
```

File: src/AmplNLWriter.jl [hunk 7]
```diff
@@ -39,6 +55,15 @@ struct _DefaultSolverCommand{F} <: AbstractSolverCommand
     f::F
 end
 
+@static if VERSION < v"1.8"
+    _get_blas_libs() = ""
+else
+    function _get_blas_libs()
+        config = LinearAlgebra.BLAS.lbt_get_config()
+        return join([lib.libname for lib in config.loaded_libs], ";")
```
Comment (odow) [inline src/AmplNLWriter.jl:63]:
```suggestion
        return join([rstrip(lib.libname, '\\') for lib in config.loaded_libs], ";")
```

**Comment by odow** [comment]
@staticfloat any ideas? Our `ipopt.exe` binary uses `OpenBLAS32_jll` directly and doesn't use LBT. I just get a 'could not spawn' error: https://github.com/jump-dev/AmplNLWriter.jl/actions/runs/4495954495/jobs/7910133289?pr=163#step:6:173
```
MOI.get(model, MOI.RawStatusString()) = "Error calling the solver. Failed with: Base.IOError(\"could not spawn `'C:\\\\Users\\\\runneradmin\\\\.julia\\\\artifacts\\\\774ef6a9c59cbea84ec6a86bf6d7a2dc39eff2e6\\\\bin\\\\ipopt.exe' 'C:\\\\Users\\\\RUNNER~1\\\\AppData\\\\Local\\\\Temp\\\\jl_eXBXKp\\\\model.nl' -AMPL 'option_file_name=D:\\\\a\\\\AmplNLWriter.jl\\\\AmplNLWriter.jl\\\\test\\\\ipopt.opt' print_level=1`: unknown error (UNKNOWN)\", -4094)"
```

I don't have a windows machine, so I'm a but out of luck.

We know that the LibraryProduct works fine on Windows, because that's what https://github.com/jump-dev/Ipopt.jl is. This is just a problem with the ExecutableProduct.

**Comment by staticfloat** [comment]
I don't really know, sorry.  It's weird to me that there are so many backslashes.... is there some kind of path escaping going on?

**Comment by amontoison** [comment]
@odow I recompiled MUMPS with LBT for Julia >= 1.9.
I will do the same thing with Ipopt.

I propose to not support anymore the artifacts with partial support of LBT when Julia 1.9 will be released.

**Comment by odow** [comment]
Okay. I'll wait for that release before spending more time trying to debug this

**Comment by odow** [comment]
> is there some kind of path escaping going on?

I tried removing some of the escapes. But I think it's just a printing artifact of how we're catching the exception, storing it as a string, and then printing it.

**Comment by odow** [comment]
It can't even spawn `$exe -v`, so something is massively wrong.

**Comment by amontoison** [comment]
@odow For Windows we don't need to modify something because I precompiled it with `OpenBLAS32_jll`. If the old version of Ipopt_jll 300.1400.400 is working, I don't understand why something changed with the last artifacts.
We need to provide additional arguments for LBT only if `!Sys.windows()`.

**Comment by odow** [comment]
Let's see how https://github.com/JuliaPackaging/Yggdrasil/pull/6734 goes.

File: Project.toml [hunk 8]
```diff
@@ -3,16 +3,19 @@ uuid = "7c4d4715-977e-5154-bfe0-e096adeac482"
...
 
 [deps]
-Ipopt_jll = "9cc047cb-c261-5740-88fc-0cf96f7bdcc7"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
+OpenBLAS32_jll = "656ef2d0-ae68-5445-9ca0-591084a874a2"
 
 [compat]
-Ipopt_jll = "=300.1400.400"
+Ipopt_jll = "300.1400.400"
```
Comment (amontoison) [inline Project.toml:11]:
```suggestion
Ipopt_jll = "=300.1400.1200"
```

Comment (amontoison) [inline Project.toml:11]:
```suggestion
Ipopt_jll = "=300.1400.1200"
```

**Comment by odow** [comment]
@amontoison is used the new version:
image
but what are we missing?

It was compiled with https://buildkite.com/julialang/yggdrasil/builds/2943#01880319-dbcd-47ac-bf17-436ae373a5d4
image
but it used OpenBLAS 0.3.21. Could that be a problem?

It's using MUMPS 500.500.101, which is https://github.com/JuliaPackaging/Yggdrasil/pull/5599, which uses OpenBLAS on Windows.

**Comment by amontoison** [comment]
@odow 
I found this message in the logs: https://buildkite.com/julialang/yggdrasil/builds/2943#01880319-dbcd-47ac-bf17-436ae373a5d4/638-1459

**Comment by odow** [comment]
Hoorah. Only. What does that mean.

**Comment by amontoison** [comment]
@odow 
I recompiled the old release 3.14.4, can you test with my fork https://github.com/amontoison/Ipopt_jll.jl ?
The version of the artifact is `300.1400.1066` (3.14.4 + :imp: 666)

**Comment by amontoison** [comment]
Can you add a `println(Base.BinaryPlatforms.host_triplet())`?
It seems that Ipopt 3.14.4 is not working with the updated dependencies.
The culprit is OpenBLAS32 or MUMPS_seq.

**Comment by odow** [comment]
`x86_64-w64-mingw32-libgfortran5-cxx11-julia_version+1.9.0`

It pulled in Julia v1.9.0?

File: test/MOI_wrapper.jl [hunk 9]
```diff
@@ -12,6 +12,19 @@ import AmplNLWriter
...
+        lbt_default_libs = get(ENV, "LBT_DEFAULT_LIBS", AmplNLWriter._get_blas_libs())
+        if isempty(lbt_default_libs)
+            run(`$exe -v`)
+        else
+            run(addenv(`$exe -v`, "LBT_DEFAULT_LIBS" => lbt_default_libs))
+        end
+    end
+    path() do exe
+        run(`$exe -v`)
+    end
```
Comment (odow) [inline test/MOI_wrapper.jl:25]: @amontoison I think you need some tests like this.

In the normal tests, we can't open the executable, but we trap the error and return `OTHER_ERROR` to MOI so it looks as if the binary didn't find a solution, but really it crashed.

**Comment by odow** [comment]
@amontoison the windows executable is still a problem

**Comment by amontoison** [comment]
I don't know where is the issue Oscar. :(
I should try to compile with an older `CompilerSupportLibraries_jll`.

**Comment by giordano** [comment]
That doesn't make much sense since we do dynamic linking. But using an older gcc+binutils can make a difference

**Comment by amontoison** [comment]
> That doesn't make much sense since we do dynamic linking.

Indeed, it is not relevant.

> But using an older gcc+binutils can make a difference

The issue is that ipopt uses libgfortran and we can't use a compiler older than GCC8 if we link with `libgfortran5`.
What I don't understand is that is that we have now an error if we recompile the working version `300.1400.400` with exactly the same options.

**Comment by amontoison** [comment]
@odow 
https://buildkite.com/julialang/yggdrasil/builds/6557#018bc4c1-67d1-4f79-87e8-23858206da1a/614-1276

**Comment by amontoison** [comment]
@odow Can you try the last `Ipopt_jll.jl` ?

File: test/MINLPTests/Project.toml [hunk 10]
```diff
@@ -10,7 +10,7 @@ Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 [compat]
 Bonmin_jll = "100.800.801"
 Couenne_jll = "0.500.801"
-Ipopt_jll = "=300.1400.400"
+Ipopt_jll = "300.1400.400, 300.1400.1300"
```
Comment (odow) [inline test/MINLPTests/Project.toml:13]:
```suggestion
Ipopt_jll = "300.1400.400, 300.1400.1400"
```

File: Project.toml [hunk 11]
```diff
@@ -3,16 +3,20 @@ uuid = "7c4d4715-977e-5154-bfe0-e096adeac482"
...
 [deps]
-Ipopt_jll = "9cc047cb-c261-5740-88fc-0cf96f7bdcc7"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
+OpenBLAS32_jll = "656ef2d0-ae68-5445-9ca0-591084a874a2"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
 
 [compat]
-Ipopt_jll = "=300.1400.400"
+Ipopt_jll = "300.1400.400, 300.1400.1300"
```
Comment (odow) [inline Project.toml:12]:
```suggestion
Ipopt_jll = "300.1400.400, 300.1400.1400"
```

**Comment by odow** [comment]
Fixed at last

**Comment by odow** [comment]
image

**Comment by amontoison** [comment]
Amazing @odow 🎉 🎉 
What was the issue finally?

**Comment by odow** [comment]
No idea. But you must have done something. Perhaps the openmp stuff? Or just general Julia/lbt improvements.

**Pull request merged**